File: src-tauri/src/shadow_server/planning/ports/saved_weekplan_reader.rs
//! Read port for Saved Weekplan ownership and body (used by consolidated list writes).

use crate::shadow_server::{
    planning::models::WeekPlanV1,
    platform::RepoError,
    shopping_list::models::SavedConsolidatedShoppingListRecord,
};

/// Weekplan context required to save a consolidated shopping list.
pub struct SavedWeekplanConsolidatedContext {
    pub body: WeekPlanV1,
    pub existing_list: Option<SavedConsolidatedShoppingListRecord>,
}

/// Planning read port for consolidated shopping list operations.
pub trait SavedWeekplanReader: Send + Sync {
    /// Loads weekplan body and any existing consolidated list after principal scoping.
    fn get_for_consolidated_list_ops(
        &self,
        plan_id: &str,
        user_id: &str,
    ) -> Result<SavedWeekplanConsolidatedContext, RepoError>;
}

File: server/services/planning/ports/savedWeekplanReader.ts
import type { WeekPlanV1 } from '../../../../types/planning'
import type { AppDb } from '../../../db/sqlite'
import type { SavedConsolidatedShoppingListRecord } from '../../shopping-list/consolidatedShoppingListRepository'
import type { PlanningPrincipal } from '../planningPrincipal'
import type { PlanningResult } from '../planningResult'

/** Weekplan context required for consolidated shopping list reads and writes. */
export interface SavedWeekplanConsolidatedContext {
  body: WeekPlanV1
  existingList: SavedConsolidatedShoppingListRecord | null
}

/** Principal-scoped weekplan row used by Planning CRUD and read-model assembly. */
export interface SavedWeekplanAccessRow {
  id: string
  name: string
  body: WeekPlanV1
  createdAt: string
  updatedAt: string
  consolidatedShoppingList: SavedConsolidatedShoppingListRecord | null
}

/** Planning read port for Saved Weekplan ownership and body (consumer: Planning + Shopping list). */
export interface SavedWeekplanReader {
  getForConsolidatedListOps(
    db: AppDb,
    planId: string,
    principal: PlanningPrincipal,
  ): Promise<PlanningResult<SavedWeekplanConsolidatedContext>>

  getById(
    db: AppDb,
    planId: string,
    principal: PlanningPrincipal,
  ): Promise<PlanningResult<SavedWeekplanAccessRow>>
}
